go/glome/glome.go

```diff
@@ -0,0 +1,223 @@
+	ErrInvalidPublicKey = fmt.Errorf("Invalid public key - byte slice len is not %d", PublicKeySize)
+	// ErrInvalidPrivateKey denotes that a slice that intend to be a private key is not of desired length
+	ErrInvalidPrivateKey = fmt.Errorf("Invalid private key - byte slice len is not %d", PrivateKeySize)
+	// ErrInvalidTagSize denotes that provided integer is not suitable to be MinPeerTagSize
+	ErrInvalidTagSize = fmt.Errorf("Invalid tag size - MinPeerTagSize must be in range %d-%d",
+		MinTagSize, MaxTagSize)
+)
+
+// PublicKey is the type of GLOME public Keys.
+//
+// It can be initializes either by casting a [PublicKeySize]byte array or from a byte
+// slice with the PublicKeyFromSlice function. For example:
+//
+//     //Generate Private Key as existing byte array
+//     b := [32]byte{0,2,...,7,6}
+//     p := glome.PublicKey(b)
+//     //Generate from byte slice
+//     s := b[:]
+//     p, _ := glome.PublicKeyFromSlice(s)
+type PublicKey [PublicKeySize]byte
+
+// PublicKeyFromSlice generates a PublicKey object from slice. Return ErrInvalidPublicKey
+// if slice's length is not PublicKeySize.
+func PublicKeyFromSlice(b []byte) (*PublicKey, error) {
+	if len(b) != PublicKeySize {
+		return nil, ErrInvalidPublicKey
+	}
+
+	var p PublicKey
+	copy(p[:], b)
+	return &p, nil
+}
+
+// PrivateKey is the type of GLOME public Keys.
+//
+// It can be initializes either by casting a [PrivateKeySize]byte array or from a byte
+// slice with the PrivateKeyFromSlice function. For example:
+//
+//     //Generate Private Key as existing byte array
+//     b := [32]byte{0,2,...,7,6}
+//     p := glome.PrivateKey(b)
+//     //Generate from byte slice
+//     s := b[:]
+//     p, _ := glome.PrivateKeyFromSlice(s)
+type PrivateKey [PrivateKeySize]byte
+
+// Public returns the PublicKey corresponding to priv.
+func (priv *PrivateKey) Public() (*PublicKey, error) {
+	slice, err := curve25519.X25519(priv[:], curve25519.Basepoint)
+	if err != nil {
+		return nil, err
+	}
+
+	p, _ := PublicKeyFromSlice(slice)
+	return p, nil
+}
+
+// Exchange generates a Dialog struct. It performs GLOME handshake, and stores create
+// a Dialog from the user to the peer. Sets MinPeerTagSize as MaxTagSize.
+func (priv *PrivateKey) Exchange(peer *PublicKey) (*Dialog, error) {
+	s, err := curve25519.X25519(priv[:], peer[:])
+	if err != nil {
+		return nil, err
+	}
+
+	public, err := priv.Public()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Dialog{shared: s, User: *public, Peer: *peer, MinPeerTagSize: MaxTagSize}, nil
+}
+
+// TruncatedExchange generates a Dialog struct. It performs GLOME handshake,
+// and stores create a Dialog from the user to the peer. Sets param m as MinPeerTagSize.
+func (priv *PrivateKey) TruncatedExchange(peer *PublicKey, m uint) (*Dialog, error) {
+	if m == 0 || m > MaxTagSize {
+		return nil, ErrInvalidTagSize
+	}
+
+	d, err := priv.Exchange(peer)
+	if err != nil {
+		return nil, err
+	}
+
+	d.MinPeerTagSize = m
```

Comment:
You validate the field but you make it exported. That means I could go and do `d.MinPeerTagSize = 0` later. I think it'd be better to leave it unexported.

Reply:
Left it unexported.